tests/candev/main.c

```diff
@@ -0,0 +1,213 @@
+        for (int i = 0; i < can_dlc; i++) {
+            printf("0x%X ", buf[i]);
+        }
+        puts("");
+    }
+
+    return 0;
+}
+
+static const shell_command_t shell_commands[] = {
+    { "send", "send some data", _send },
+    { "receive", "receive some data", _receive },
+    { NULL, NULL, NULL }
+};
+
+static void _can_event_callback(candev_t *dev, candev_event_t event, void *arg)
+{
+    (void)arg;
+    struct can_frame *frame;
+
+    switch (event) {
+        case CANDEV_EVENT_ISR:
+            DEBUG("_can_event: CANDEV_EVENT_ISR\n");
+            dev->driver->isr(candev);
+            break;
+        case CANDEV_EVENT_WAKE_UP:
+            DEBUG("_can_event: CANDEV_EVENT_WAKE_UP\n");
+            break;
+        case CANDEV_EVENT_TX_CONFIRMATION:
+            DEBUG("_can_event: CANDEV_EVENT_TX_CONFIRMATION\n");
+            break;
+        case CANDEV_EVENT_TX_ERROR:
+            DEBUG("_can_event: CANDEV_EVENT_TX_ERROR\n");
+            break;
+        case CANDEV_EVENT_RX_INDICATION:
+            DEBUG("_can_event: CANDEV_EVENT_RX_INDICATION\n");
+
+            frame = (struct can_frame *)arg;
+
+            DEBUG("\tid: %" PRIx32 " dlc: %" PRIx8 " Data: \n\t", frame->can_id,
+                  frame->can_dlc);
+            for (uint8_t i = 0; i < frame->can_dlc; i++) {
+                DEBUG("0x%X ", frame->data[i]);
+            }
+            DEBUG(" ");
+
+            //Store in buffer until user requests the data
+            isrpipe_write_one(&rxbuf,
+                              (uint8_t)((frame->can_id & 0x1FFFFFFF) >> 24));        //exclude flags
+            isrpipe_write_one(&rxbuf,
+                              (uint8_t)((frame->can_id & 0xFF0000) >> 16));
+            isrpipe_write_one(&rxbuf, (uint8_t)((frame->can_id & 0xFF00) >> 8));
+            isrpipe_write_one(&rxbuf, (uint8_t)((frame->can_id & 0xFF)));
+
+            isrpipe_write_one(&rxbuf, frame->can_dlc);
+            for (uint8_t i = 0; i < frame->can_dlc; i++) {
+                isrpipe_write_one(&rxbuf, frame->data[i]);
+            }
+
+            break;
+        case CANDEV_EVENT_RX_ERROR:
+            DEBUG("_can_event: CANDEV_EVENT_RX_ERROR\n");
+            break;
+        case CANDEV_EVENT_BUS_OFF:
+            dev->state = CAN_STATE_BUS_OFF;
+            break;
+        case CANDEV_EVENT_ERROR_PASSIVE:
+            dev->state = CAN_STATE_ERROR_PASSIVE;
+            break;
+        case CANDEV_EVENT_ERROR_WARNING:
+            dev->state = CAN_STATE_ERROR_WARNING;
+            break;
+        default:
+            DEBUG("_can_event: unknown event\n");
+            break;
+    }
+}
+
+int main(void)
+{
+    uint8_t rx_ringbuf[RX_RINGBUFFER_SIZE] = { 0 };
+    (void) _can_event_callback;
+
+    puts("candev test application\n");
+
+    isrpipe_init(&rxbuf, (uint8_t *)rx_ringbuf, sizeof(rx_ringbuf));
+#ifdef BOARD_NATIVE
+    puts("Initializing Linux Can device");
+    candev_linux_init( &linux_dev, &(candev_linux_conf[0]));    //vcan0
```

Comment:
Is `auto_init_can` not taking care of this?

Reply:
no, because `auto_init_can` is not used here. The whole point of this test-app is to used the candev abstraction without using the can-dll.